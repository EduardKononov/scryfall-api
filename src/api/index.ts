export { default as Cards } from './card';
export * from './card';
export { default as CardIdentifierBuilder } from './card-identifier';
export * from './card-identifier';
export { default as Catalogs } from './catalog';
export * from './catalog';
export { default as Rulings } from './ruling';
export * from './ruling';
export { default as Sets } from './set';
export * from './set';
export { default as SymbologyApi } from './symbology';
export * from './symbology';
