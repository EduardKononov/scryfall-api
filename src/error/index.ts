export { default as InvalidScryfallArgumentError } from './invalid.error';
export { default as ScryfallError } from './scryfall.error';
export { default as UnknownScryfallError } from './unknown.error';
