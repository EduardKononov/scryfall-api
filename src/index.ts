export * from './api/CardIdentifierBuilder';
export { default as Cards } from './api/Cards';
export { default as Catalog } from './api/Catalog';
export { default as Rulings } from './api/Rulings';
export { default as Sets } from './api/Sets';
export { default as Symbology } from './api/Symbology';
export * from './error/MagicQueryError';
export * from './query/MagicPageResult';
export * from './query/MagicQuery';
export * from './types';
