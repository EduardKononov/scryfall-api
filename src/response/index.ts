export * from './error.response';
export * from './list.response';
