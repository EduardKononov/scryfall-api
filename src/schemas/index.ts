export { default as CardSymbol } from './card-symbol';
export { default as Color } from './color';
export { default as ErrorResponse } from './error.response';
export { BaseSchema, default as ListResponse } from './list.response';
export { default as ManaCost } from './mana-cost';
