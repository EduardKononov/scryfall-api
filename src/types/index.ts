export * from './Border';
export * from './BulkDataDefinition';
export * from './BulkDataType';
export * from './BulkDataTypes';
export * from './Card';
export * from './CardFace';
export * from './CardIdentifier';
export * from './CardSymbol';
export * from './Color';
export * from './Format';
export * from './FrameEffect';
export * from './Game';
export * from './HomepageLink';
export * from './ImageUris';
export * from './Layout';
export * from './Legalties';
export * from './Legalty';
export * from './ManaCost';
export * from './PageQuery';
export * from './PageQueryValue';
export * from './Preview';
export * from './Prices';
export * from './PromoType';
export * from './PurchaseUris';
export * from './Query';
export * from './Rarity';
export * from './RelatedCard';
export * from './RelatedCardComponent';
export * from './RelatedUris';
export * from './ResultCatalog';
export * from './ResultData';
export * from './ResultList';
export * from './RetryStrategy';
export * from './Ruling';
export * from './SearchError';
export * from './SearchOptions';
export * from './Set';
export * from './SetType';
export * from './Sort';
export * from './SortDirection';
export * from './TOrArrayOfT';
export * from './UniqueStrategy';
