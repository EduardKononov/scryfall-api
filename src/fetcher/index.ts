import { ScryfallError } from '../error';
import type { TOrArrayOfT } from '../types';
import createDebounceFetcher from './debounce.fetcher';
import type { FetcherParams } from './fetcher.params';
import MagicPageResult from './page.result';
import createRetryFetcher from './retry.fetcher';

const simpleFetcher = createDebounceFetcher(createRetryFetcher());

const endpoint = 'https://api.scryfall.com';

export default async function fetcher<TData>(
	apiPath: TOrArrayOfT<number | string | undefined>,
	params?: FetcherParams,
	body?: object
): Promise<TData | undefined> {
	let path: number | string = '';
	if (typeof apiPath === 'number' || typeof apiPath === 'string') {
		path = apiPath;
	} else if (apiPath) {
		path = apiPath.join('/');
	}

	const url = new URL(`${endpoint}/${path}`);
	if (params) {
		Object.entries(params).forEach(([key, value]) => {
			if (value !== undefined) {
				url.searchParams.append(key, `${value}`);
			}
		});
	}

	try {
		const init: RequestInit = {
			method: body ? 'POST' : 'GET',
		};
		if (body) {
			init.body = JSON.stringify(body);
			init.headers = {
				'Content-Type': 'application/json',
			};
		}
		const response = await simpleFetcher(url as URL, init);

		return (await response.json()) as TData;
	} catch (e) {
		if (e instanceof ScryfallError) {
			if (e.warnings) {
				console.warn(e.warnings);
			}
			if (e.code === 'not_found' || e.code === 'bad_request') {
				return undefined;
			}
		}
		throw e;
	}
}

export { MagicPageResult };
