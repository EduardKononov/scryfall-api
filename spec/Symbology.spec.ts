import { describe, expect, it } from 'bun:test';

import { InvalidScryfallArgumentError, Symbology } from '../src';

describe('Scryfall', () => {
	describe('Symbology', () => {
		describe('.all()', () => {
			it('should works as expected', async () => {
				const symbology = await Symbology.all();
				expect(symbology.length).toBeGreaterThanOrEqual(75);
			});
		});
		describe('.parseMana()', () => {
			it('should works as expected', async () => {
				const cost = await Symbology.parseMana('2ww');
				expect(cost).toMatchSnapshot();
			});
			it('should throw if invalid shorthand', () => {
				// @ts-expect-error test
				expect(() => Symbology.parseMana()).toThrow(InvalidScryfallArgumentError);
				// @ts-expect-error test
				expect(() => Symbology.parseMana(1)).toThrow(InvalidScryfallArgumentError);
				// @ts-expect-error test
				expect(() => Symbology.parseMana('')).toThrow(InvalidScryfallArgumentError);
			});
		});
	});
});
