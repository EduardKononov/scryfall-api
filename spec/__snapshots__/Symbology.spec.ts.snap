// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Scryfall Symbology .parseMana() should throw if empty shorthand 1`] = `[Function]`;
